src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
     pub mod line {
         use super::*;
         ensogl::define_shape_system! {
-            () {
+            (focus_split_center:Vector2<f32>, focus_split_angle:f32) {
                 let width  = LINE_WIDTH.px();
                 let height : Var<Pixels> = "input_size.y".into();
-                let shape  = Rect((width,height));
-                let shape  = shape.fill(color::Rgba::from($($color)*));
-                shape.into()
+                let shape  = Rect((width.clone(),height));
+
+                let split_shape = FocusedEdge::new(
+                    shape,&focus_split_center.px(),&focus_split_angle.into());
+                let shape       = split_shape.fill($color, $highlight_color);
+                hover_area(shape,HOVER_EXTENSION.px()).into()
+            }
+        }
+
+        impl EdgeShape for component::ShapeView<Shape> {
+            fn set_focus_split_center_local(&self, center:Vector2<f32>) {
+                self.shape.focus_split_center.set(center);
+            }
+
+            fn set_focus_split_angle(&self, angle:f32) {
+                self.shape.focus_split_angle.set(angle);
+            }
+
+            fn sprite(&self) -> &Sprite{
+                &self.shape.sprite
+            }
+
+            fn events(&self) -> &ShapeViewEvents{
+                &self.events
+            }
+
+            fn normal_local(&self, _:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+                nalgebra::Rotation2::new(0.0)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                #[allow(deprecated)]
+                let height = self.sprite().size.get().y;
+                let y      = point.y.clamp(-height/2.0, height/2.0);
+                Some(Vector2::new(0.0, y))
             }
         }
     }
 }}
 
-macro_rules! define_arrow {($($color:tt)*) => {
+macro_rules! define_arrow {($color:expr, $highlight_color:expr) => {
     /// Shape definition.
     pub mod arrow {
         use super::*;
         ensogl::define_shape_system! {
-            () {
+            (focus_split_center:Vector2<f32>, focus_split_angle:f32) {
                 let width  : Var<Pixels> = "input_size.x".into();
                 let height : Var<Pixels> = "input_size.y".into();
-                let width      = width  - (2.0 * PADDING).px();
-                let height     = height - (2.0 * PADDING).px();
-                let triangle   = Triangle(width,height);
-                let offset     = (LINE_WIDTH/2.0).px();
-                let triangle_l = triangle.translate_x(-&offset);
-                let triangle_r = triangle.translate_x(&offset);
-                let shape      = triangle_l + triangle_r;
-                let shape      = shape.fill(color::Rgba::from($($color)*));
+                let focus_split_angle  = focus_split_angle.into();
+                let focus_split_center = focus_split_center.px();
+                let shape = Triangle(width-1.0.px(),height-1.0.px());
```

Comment:
what are these `-1.px()` ? Btw you can write `1.px()` instead of `1.0.px()`. They should be named variable telling what they are for.